File: Algorithms/p017_Letter_Combinations_of_a_Phone_Number/p017_Letter_Combinations_of_a_Phone_Number.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if not digits:
            return []
        for c in digits:
            if c not in map(lambda x:str(x),range(0,10)):
                return []
        res = []
        mapping = [' ',
                   '*',
                   'abc',
                   'def',
                   'ghi',
                   'jkl',
                   'mno',
                   'pqrs',
                   'tuv',
                   'wxyz']
        cur = []
        self.dfs(res,mapping,cur,digits,0)
        return res

    def dfs(self, res, mapping, cur, digits, pos):
        if pos<len(digits):
            d = int(digits[pos])
            for i in range(0,len(mapping[d])):
                cur.append(mapping[d][i])
                self.dfs(res,mapping,cur,digits,pos+1)
                cur.pop()
        else:
            res.append(''.join(cur))

File: Algorithms/p017_Letter_Combinations_of_a_Phone_Number/p017_Letter_Combinations_of_a_Phone_Number.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if(digits.empty()) return {};
        for(char c : digits){
            if(c<'2'||c>'9') return {};
        }
        vector<string> res;
        vector<vector<char>> mapping = {{'a','b','c'},
                                        {'d','e','f'},
                                        {'g','h','i'},
                                        {'j','k','l'},
                                        {'m','n','o'},
                                        {'p','q','r','s'},
                                        {'t','u','v'},
                                        {'w','x','y','z'}};
        string cur;
        dfs(res,mapping,cur,digits,0);
        return res;
    }
    void dfs(vector<string>& res, vector<vector<char>>& mapping, string& cur, string& digits, int pos) {
        if(pos<digits.size()){
            int d = digits[pos]-'0'-2;
            for(int i=0;i<mapping[d].size();++i){
                cur.push_back(mapping[d][i]);
                dfs(res,mapping,cur,digits,pos+1);
                cur.pop_back();
            }
        }else{
            res.push_back(cur);
        }
    }
};
